crates/bevy_ecs/src/relationship/mod.rs

```diff
@@ -0,0 +1,262 @@
+mod relationship_query;
+mod relationship_source_collection;
+
+pub use related_methods::*;
+pub use relationship_query::*;
+pub use relationship_source_collection::*;
+
+pub use bevy_ecs_macros::{Relationship, RelationshipSources};
+
+use crate::{
+    component::{Component, ComponentId, Mutable},
+    entity::Entity,
+    system::{
+        command::HandleError,
+        entity_command::{self, CommandWithEntity},
+        error_handler,
+    },
+    world::DeferredWorld,
+};
+use log::warn;
+
+/// A [`Component`] on a "source" [`Entity`] that references another target [`Entity`], creating a "relationship" between them. Every [`Relationship`]
+/// has a corresponding [`RelationshipSources`] type (and vice-versa), which exists on the "target" entity of a relationship and contains the list of all
+/// "source" entities that relate to the given "target"
+///
+/// The [`Relationship`] component is the "source of truth" and the [`RelationshipSources`] component reflects that source of truth. When a [`Relationship`]
+/// component is inserted on an [`Entity`], the corresponding [`RelationshipSources`] component is immediately inserted on the target component if it does
+/// not already exist, and the "source" entity is automatically added to the [`RelationshipSources`] collection (this is done via "component hooks").
+///
+/// A common example of a [`Relationship`] is the parent / child relationship. Bevy ECS includes a canonical form of this via the [`Parent`](crate::hierarchy::Parent)
+/// [`Relationship`] and the [`Children`](crate::hierarchy::Children) [`RelationshipSources`].
+///
+/// [`Relationship`] and [`RelationshipSources`] should always be derived to ensure the hooks are set up properly. They will both automatically
+/// implement [`Component`] with the necessary configuration to drive the [`Relationship`].
+///
+/// ```
+/// # use bevy_ecs::relationship::{Relationship, RelationshipSources};
+/// # use bevy_ecs::entity::Entity;
+/// #[derive(Relationship)]
+/// #[relationship(relationship_sources = Children)]
+/// pub struct Parent(pub Entity);
+///
+/// #[derive(RelationshipSources)]
+/// #[relationship_sources(relationship = Parent)]
+/// pub struct Children(Vec<Entity>);
+/// ```
+///
+/// When deriving [`RelationshipSources`] you can specify the `#[relationship_sources(despawn_descendants)]` attribute to
+/// automatically despawn entities stored in an entity's [`RelationshipSources`] when that entity is despawned:
+///
+/// ```
+/// # use bevy_ecs::relationship::{Relationship, RelationshipSources};
+/// # use bevy_ecs::entity::Entity;
+/// #[derive(Relationship)]
+/// #[relationship(relationship_sources = Children)]
+/// pub struct Parent(pub Entity);
+///
+/// #[derive(RelationshipSources)]
+/// #[relationship_sources(relationship = Parent, despawn_descendants)]
+/// pub struct Children(Vec<Entity>);
+/// ```
+///
+// NOTE: The "deprecated" state is used to prevent users from mutating the internal RelationshipSource collection.
```

Comment:
Mentioned elsewhere, but I much prefer the "scary name" approach over the "permanently deprecated" strategy :)